### PR TITLE
gh-115911: Ignore PermissionError during import from cwd

### LINKED ISSUE #115911: importlib: PermissionError during startup if working directory isn't readable
# Bug report

### Bug description:

On macOS `importlib._bootstrap_external.PathFinder._path_importer_cache()` raises `PermissionError` during interpreter startup if `''` is included in `sys.path` and the current working directory is not readable.

# Reproduction

Given a CWD that is not readable by fred, the user fred cannot run Python
```
➜  private su fred -c "whoami; python3.13 -c 'pass'"
Password:
fred
Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1322, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1262, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1544, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1516, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1495, in _path_importer_cache
PermissionError: [Errno 13] Permission denied

➜  private uname -mv; whoami; pwd; ls -ld                                                        
Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000 arm64
alex
/Users/alex/private
drwx------  2 alex  staff  64 25 Feb 10:44 .

➜  private python3.13 -c "import sys;print(sys.version)"           
3.13.0a4+ (heads/main:6550b54813, Feb 25 2024, 10:56:11) [Clang 15.0.0 (clang-1500.1.0.2.5)]
```
# Workaround
Adding `-P`, prevents `''` being added to `sys.path`, so avoids the exception
```
➜  private su fred -c "whoami; python3.13 -P -c 'pass'"                                    
Password:
fred
```

# Discussion

On macOS the libc function `getcwd()` can return `EACCES`. From the manpage

> [EACCES] Read or search permission was denied for a component of the pathname.  This is only checked in limited cases, depending on implementation details.

When searching for importable modules `PathFinder._path_importer_cache()` attempts to determine the cwd by calling `os.getcwd()`, it handles a `FileNotFoundError` exception, but not `PermissionError`. Because `PathFinder` is used during interpreter startup user code has no opportunity to catch the exception.

# Proposed fix

Ignore `PermissionError` in `PathFinder._path_importer_cache()`, the same way `FileNotFoundError` is currently. This would result in imports succeeding, but without getting cached. E.g. applying the below change & rebuilding/reinstalling

```
➜  private su fred -c "whoami; python3.13 -c 'import sys;print(sys.version)'"
Password:
fred
3.13.0a4+ (heads/main:6550b54813, Feb 25 2024, 10:56:11) [Clang 15.0.0 (clang-1500.1.0.2.5)]
```

I'm happy to submit a PR with this, and unit tests as deemed suitable

```diff
➜  cpython git:(main) ✗ git diff
diff --git a/Lib/importlib/_bootstrap_external.py b/Lib/importlib/_bootstrap_external.py
index 2a9aef0317..5d1f4f1de0 100644
--- a/Lib/importlib/_bootstrap_external.py
+++ b/Lib/importlib/_bootstrap_external.py
@@ -1493,7 +1493,7 @@ def _path_importer_cache(cls, path):
         if path == '':
             try:
                 path = _os.getcwd()
-            except FileNotFoundError:
+            except (FileNotFoundError, PermissionError):
                 # Don't cache the failure as the cwd can easily change to
                 # a valid directory later on.
                 return None
```

# Other Python Versions

In Python 3.10, 3.11 & 3.12 the same exception can occur when user code imports a non-builtin module (e.g. `base64`, zlib), but it does not occur during interpreter startup. I presume this is due to a change in which modules are required during interpreter initialisation.

```
➜  private su fred -c "whoami; python3.10 -c 'import sys;print(sys.version);import zlib'"              
Password:
fred
3.10.13 (main, Aug 24 2023, 12:59:26) [Clang 15.0.0 (clang-1500.1.0.2.5)]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1002, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 945, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1439, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1408, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1366, in _path_importer_cache
PermissionError: [Errno 13] Permission denied

➜  private su fred -c "whoami; python3.11 -c 'import sys;print(sys.version);import zlib'"
Password:
fred
3.11.7 (main, Dec  4 2023, 18:10:11) [Clang 15.0.0 (clang-1500.1.0.2.5)]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1138, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1078, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1504, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1473, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1431, in _path_importer_cache
PermissionError: [Errno 13] Permission denied

➜  private su fred -c "whoami; python3.12 -c 'import sys;print(sys.version);import zlib'"
Password:
fred
3.12.1 (main, Dec  7 2023, 20:45:44) [Clang 15.0.0 (clang-1500.1.0.2.5)]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1322, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1262, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1524, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1496, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1475, in _path_importer_cache
PermissionError: [Errno 13] Permission denied
```

### CPython versions tested on:

3.10, 3.11, 3.12, CPython main branch

### Operating systems tested on:

macOS

### Linked PRs
* gh-116131

### PR DESCRIPTION
**Pull request by moreati**
On macOS `getcwd(3)` can return EACCES if a path component isn't readable, resulting in PermissionError. `PathFinder.find_spec()` now catches these and ignores them - the same treatment as a missing/deleted cwd.

Introduces `test.support.os_helper.save_mode(path, ...)`, a context manager that restores the mode of a path on exit.

This is allows finer control of exception handling and robust environment restoration across platforms in `FinderTests.test_permission_error_cwd()`.

Fixes #115911

* Issue: gh-115911

----
📚 Documentation preview 📚: https://cpython-previews--116131.org.readthedocs.build/

### CODE REVIEW

**Comment by moreati** [comment]
The new test fails (as intended) without the accompanying fix. Local test run on macOS 14.3.1
```
======================================================================
ERROR: test_permission_error_cwd (test.test_importlib.import_.test_path.Frozen_FindModuleTests.test_permission_error_cwd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/src/cpython/Lib/test/test_importlib/import_/test_path.py", line 175, in test_permission_error_cwd
    self.assertIsNone(self.machinery.PathFinder.find_spec('whatever'))
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 1544, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1516, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1495, in _path_importer_cache
PermissionError: [Errno 13] Permission denied
```

File: Lib/test/test_importlib/metadata/fixtures.py [hunk 1]
```diff
@@ -44,6 +44,16 @@ def save_cwd():
...
+@contextlib.contextmanager
+def save_mode(path, *, follow_symlinks=True):
+    path = pathlib.Path(path)
+    orig = path.stat(follow_symlinks=follow_symlinks)
+    try:
+        yield
+    finally:
+        path.chmod(orig.st_mode, follow_symlinks=follow_symlinks)
+
+
```
Comment (jaraco) [inline Lib/test/test_importlib/metadata/fixtures.py]: Tests in this path (`test_importlib/metadata`) are for the `importlib.metadata` package. Please move this to a new `test_importlib.fixtures` or another suitable location outside `metadata`. Thanks.

Comment (moreati) [reply Lib/test/test_importlib/metadata/fixtures.py]: Is it okay for `test.test_importlib.import_.test_path` to use _existing_ support functions from here? (e.g. `test.test_importlib.metadata.fixtures.save_cwd()`)

Comment (jaraco) [reply Lib/test/test_importlib/metadata/fixtures.py]: Oh. Good point. I suspect your usage is probably the first such instance (though I haven't checked).

My motivation for keeping them separate is mainly to support the backports (i.e. [`importlib_metadata`](/python/importlib_metadata)), which share the same code. Since the relationship between this code and the backports are fairly stable, I'm okay with other stdlib tests building on these fixtures. However, there may be suitable fixtures in `test.support**` (e.g. `test.support.os_helper.change_cwd` and `test.support.os_helper.temp_dir`) and those should be preferred. Can you explore using those instead?

In the meantime, I've updated https://github.com/python/cpython/issues/120801 to track consolidating the fixtures. If you do end up relying on the metadata fixtures, I'll probably end up replacing those in that effort, so if it's not too much a hassle, please consider using those. Thanks.

Comment (jaraco) [reply Lib/test/test_importlib/metadata/fixtures.py]: In #120803, I'm proposing to remove those fixtures. Please let me know if you'd like me to help adapt this PR to accommodate those changes.

Comment (moreati) [reply Lib/test/test_importlib/metadata/fixtures.py]: I'd wondered if there was appetite for consolidating such fixtures and decorators across the test suite. So will do.

Comment (jaraco) [reply Lib/test/test_importlib/metadata/fixtures.py]: Thanks!

File: Lib/test/test_importlib/import_/test_path.py [hunk 2]
```diff
@@ -153,6 +154,25 @@ def test_deleted_cwd(self):
...
+    @os_helper.skip_unless_working_chmod
+    def test_permission_error_cwd(self):
+        # gh-115911
+        with (
+            os_helper.temp_dir() as new_dir,
+            os_helper.save_mode(new_dir),
+            os_helper.save_cwd(),
+            util.import_state(path=['']),
+        ):
+            os.chdir(new_dir)
```
Comment (jaraco) [inline Lib/test/test_importlib/import_/test_path.py]: Why not re-use `change_cwd`?

```suggestion
        with (
            os_helper.temp_dir() as new_dir,
            os_helper.save_mode(new_dir),
            os_helper.change_cwd(new_dir),
            util.import_state(path=['']),
        ):
```

Consider also `save_mode` could be `change_mode` (except it's maybe not possible to implement the skip behavior).

File: Lib/test/support/os_helper.py [hunk 3]
```diff
@@ -508,9 +535,36 @@ def temp_dir(path=None, quiet=False):
             rmtree(path)
 
 
+@contextlib.contextmanager
+def save_cwd(*, quiet=False):
```
Comment (jaraco) [inline Lib/test/support/os_helper.py]: This new function may not be needed in light of `change_cwd` (see the other comment that uses it).

File: Lib/test/support/os_helper.py [hunk 4]
```diff
@@ -294,6 +294,33 @@ def skip_unless_working_chmod(test):
...
+        try:
+            os.chmod(path, saved_mode.st_mode)
+        except OSError as exc:
+            if not quiet:
+                raise
+            warnings.warn(f'tests may fail, unable to restore the mode of '
+                          f'{path!r} to {saved_mode.st_mode}: {exc}',
+                          RuntimeWarning, stacklevel=3)
+
+
```
Comment (jaraco) [inline Lib/test/support/os_helper.py]: I appreciate you thinking about `save_mode` but that leaves the caller to do half of the work (initial chmod, including error handling) but not the cleanup. Perhaps better would be a `change_mode` that (a) invokes `skip_unless_working_chmod` and (b) takes as a parameter the mode to set. This approach should also simplify the usage in the test.

```suggestion
def change_mode(test, path, mode, *, quiet=False):
    """Context manager that restores the mode (permissions) of path on exit.

    Arguments:

      test: ...
      
      path: Path of the file to restore mode of.
      
      mode: ...

      quiet: if False (the default), the context manager raises an exception
        on error.  Otherwise, it issues only a warning and keeps the current
        working directory the same.

    """
    skip_unless_working_chmod(test)
    
    saved_mode = os.stat(path)

    try:
        os.chmod(path, mode)
    except OSError as exc:
        if not quiet:
            raise
        warnings.warn(f'tests may fail, unable to change the mode '
                      f'of {path!r} to {mode!r}: {exc}',
                      RuntimeWarning, stacklevel=3)
    try:
        yield
    finally:
        os.chmod(path, saved_mode.st_mode)


```

This approach is more symmetric with `change_cwd` and saves the caller from having to deal with the consequences of the error. It may not even be valuable or necessary to handle the exception or solicit `quiet`, because the test should already have been skipped at that point.

```suggestion
def change_mode(test, path, mode):
    """Context manager that restores the mode (permissions) of path on exit.

    Arguments:

      test: ...
      
      path: Path of the file to restore mode of.
      
      mode: ...

    """
    skip_unless_working_chmod(test)
    
    saved_mode = os.stat(path)

    os.chmod(path, mode)

    try:
        yield
    finally:
        os.chmod(path, saved_mode.st_mode)


```

File: Lib/test/test_importlib/import_/test_path.py [hunk 5]
```diff
@@ -153,6 +154,25 @@ def test_deleted_cwd(self):
...
+            try:
+                os.chmod(new_dir, 0o000)
+            except OSError:
+                self.skipTest("platform does not allow "
+                              "changing mode of the cwd")
+
+            # Do not want PermissionError raised.
+            self.assertIsNone(self.machinery.PathFinder.find_spec('whatever'))
+
```
Comment (jaraco) [inline Lib/test/test_importlib/import_/test_path.py]: If accepting the `change_chmod` context:

```suggestion
    def test_permission_error_cwd(self):
        # gh-115911
        with (
            os_helper.temp_dir() as new_dir,
            os_helper.change_mode(self, new_dir), 0o000),
            os_helper.save_cwd(),
            util.import_state(path=['']),
        ):
            os.chdir(new_dir)

            # Do not want PermissionError raised.
            self.assertIsNone(self.machinery.PathFinder.find_spec('whatever'))

```

File: Doc/reference/import.rst [hunk 6]
```diff
@@ -871,10 +871,10 @@ module.
 
 The current working directory -- denoted by an empty string -- is handled
 slightly differently from other entries on :data:`sys.path`. First, if the
-current working directory is found to not exist, no value is stored in
-:data:`sys.path_importer_cache`. Second, the value for the current working
-directory is looked up fresh for each module lookup. Third, the path used for
-:data:`sys.path_importer_cache` and returned by
+current working directory cannot be determined or is found to not exist, no
```
Comment (jaraco) [inline Doc/reference/import.rst]: Nitpick - let's fix the split infinitive while we're here.

```suggestion
current working directory cannot be determined or is found not to exist, no
```

**Review by jaraco — COMMENTED** [review]
Looks good to me, but I'd like to see another review from someone more familiar with importlib internals.

File: Lib/test/test_importlib/import_/test_path.py (hunk 5 above)
Comment (moreati) [reply Lib/test/test_importlib/import_/test_path.py]: My initial design was as you suggest, but I wasn't able to make that work across all of the CI platforms and I've concluded it can't be made to work. I found that to avoid any undesired `PermissionError` during setup/teardown
1. during setup `chmod()` and `chdir()` must be called in a a precise order
1. during teardown they require another precise order
1. teardown order cannot be the reverse of setup.

Having the setup `chdir()` & setup `chmod()` calls within the context manager stack makes it impossible to satisfy item 3. Python will always unroll the context manager stack in reverse order. That is why I arrived at `save_cwd()` and `save_mode()` - so the context managers can teardown in the correct order, without imposing the wrong setup order (or vice versa).

As a specific example on macOS the order of operations must be
1. `cd testdir; chmod 000 testdir`;
2. perform test
3. `cd <saved>; chmod <saved> testdir;`

Doing otherwise during setup (chmod, cd) or teardown (chmod, cd) results in permission denied

## Wrong setup order
```
➜  tmp mkdir test1
➜  tmp chmod 000 test1
➜  tmp cd test1
cd: permission denied: test1
```
## Wrong teardown order
```
➜  tmp mkdir test2
➜  tmp cd test2
➜  test2 chmod 000 .
➜  test2 chmod u=rwx .
chmod: .: Permission denied
```
## Correct orders
```
➜  tmp mkdir test3
➜  tmp cd test3
➜  test3 chmod 000 .
➜  test3 cd ..
➜  tmp chmod u=rwx test3
➜  tmp rmdir test3 
```
```
➜  tmp sw_vers
ProductName:		macOS
ProductVersion:		14.6.1
BuildVersion:		23G93
```

Comment (moreati) [reply Lib/test/test_importlib/import_/test_path.py]: If you're satisfied with my explanation/example then I could incorporate a condensed version into the doc strings of the context managers and/or add comments in the test case.

Comment (jaraco) [reply Lib/test/test_importlib/import_/test_path.py]: Wow. That's really annoying. Thanks for tracking it all down and explaining it here. You may want to consider, in addition to a summarized or condensed explanation in the code, also link to the comment with the elaborate explanation. I'm satisfied regardless.

Comment (moreati) [reply Lib/test/test_importlib/import_/test_path.py]: Thank you, and will do.

Comment (moreati) [reply Lib/test/test_importlib/import_/test_path.py]: > You may want to consider, in addition to a summarized or condensed explanation in the code, also link to the comment with the elaborate explanation.

Done

> I'm satisfied regardless.

Thank you, I'll mark these conversations as resolved.

File: Lib/test/support/os_helper.py [hunk 7]
```diff
@@ -294,6 +294,33 @@ def skip_unless_working_chmod(test):
     return test if ok else unittest.skip(msg)(test)
 
 
+@contextlib.contextmanager
+def save_mode(path, *, quiet=False):
+    """Context manager that restores the mode (permissions) of path on exit.
```
Comment (brettcannon) [inline Lib/test/support/os_helper.py]:
```suggestion
    """Context manager that restores the mode (permissions) of *path* on exit.
```

File: Lib/test/support/os_helper.py [hunk 8]
```diff
@@ -294,6 +294,33 @@ def skip_unless_working_chmod(test):
...
+
+    Arguments:
+
+      path: Path of the file to restore mode of.
```
Comment (brettcannon) [inline Lib/test/support/os_helper.py]:
```suggestion
      path: Path of the file to restore the mode of.
```

File: Lib/test/support/os_helper.py [hunk 9]
```diff
@@ -508,9 +535,38 @@ def temp_dir(path=None, quiet=False):
             rmtree(path)
 
 
+@contextlib.contextmanager
+def save_cwd(*, quiet=False):
+    """Return a context manager that restores the current working directory on
+    exit.
+
+    Prefer change_cwd() if you can. For most use cases it is cleaner.
```
Comment (brettcannon) [inline Lib/test/support/os_helper.py]: If this is true then why have this? Couldn't you make `change_cwd()` not take an argument and thus make it use the current working directory as what to change to as a way to forcibly restore back to original cwd?

Otherwise can you just call `change_cwd()` so this is just wrapping that?

File: Lib/test/test_importlib/import_/test_path.py [hunk 10]
```diff
@@ -153,6 +154,29 @@ def test_deleted_cwd(self):
...
+    def test_permission_error_cwd(self):
+        # gh-115911: Test that an unreadable CWD does not break imports, in
+        # particular during early stages of interpreter startup.
+        with (
+            os_helper.temp_dir() as new_dir,
+            os_helper.save_mode(new_dir),
+            os_helper.save_cwd(),
+            util.import_state(path=['']),
+        ):
+            # chdir() & chmod() are done here (inside the with block) because
```
Comment (brettcannon) [inline Lib/test/test_importlib/import_/test_path.py]:
```suggestion
            # chdir() & chmod() are done here (inside the 'with' block) because
```

**Review by brettcannon — DISMISSED** [review]
I think one of the decorators you're adding might not be necessary, or at least can be simplified. Otherwise LGTM!

File: Lib/test/support/os_helper.py (hunk 9 above)
Comment (moreati) [reply Lib/test/support/os_helper.py]: I confess I can't remember why I concluded `save_cwd()` was necessary for `test_permission_error_cwd()` (maybe a corner case on Windows?). So I've kicked off https://github.com/python/cpython/pull/124091 to shed light either way.

If #124091 shows `save_cwd()` isn't needed, then I'll remove it

Otherwise, I'll reply again with some clarifying questions. Thanks for the kick.

Comment (moreati) [reply Lib/test/support/os_helper.py]: Result: All tests passed on all platforms after deleting `save_cwd()` and using `change_cwd()` in https://github.com/python/cpython/actions/runs/10860873897.
Conclusion: `save_cwd()` is unnecessary for this PR. I'll remove it.

**Comment by moreati** [comment]
I have made the requested changes; please review again

**Comment by moreati** [comment]
@brettcannon I believe I've made the changes you requested. Please could you take another look?

**Review by jaraco — APPROVED** [review]
Given that you address Brett's changes, I'm inclined to push this forward. We can revisit (and revert) if necessary.

**Comment by moreati** [comment]
I *think* the buildbot failure is a flaky test, atleast on that runner. Of the 5 previous builds 3 failed on this test ([3138](https://buildbot.python.org/#/builders/1366/builds/3138), [3139](https://buildbot.python.org/#/builders/1366/builds/3139), [3140](https://buildbot.python.org/#/builders/1366/builds/3140)) have failed in the same test, 2 succeeded ([3142](https://buildbot.python.org/#/builders/1366/builds/3142), [3143](https://buildbot.python.org/#/builders/1366/builds/3143)).

Trying to reproduce the failure locally I saw it with the revision from this PR (a8dc6d6d44a141a8f839deb248a02148dcfb509e), but not previous revision in main (914c232e939).

**Comment by moreati** [comment]
A second local run with rev a8dc6d6 has succeeded

```
...
    test_launcher test_lzma test_msvcrt test_perf_profiler
    test_perfmaps test_smtpnet test_ssl test_startfile test_tcl
    test_tkinter test_ttk test_ttk_textonly test_turtle test_winapi
    test_winconsoleio test_winreg test_winsound test_wmi

1 test skipped (resource denied):
    test_zipfile64

430 tests OK.

Total duration: 17 min 8 sec
Total tests: run=43,007 skipped=2,302
Total test files: run=483/484 skipped=53 resource_denied=1
Result: SUCCESS
➜  cpython git:(main) ✗ git rev-parse HEAD   
a8dc6d6d44a141a8f839deb248a02148dcfb509e
```
